components/ui/PortalPopup.tsx

```diff
@@ -0,0 +1,228 @@
+    switch (placement) {
+      case PopupPlacement.Bottom:
+      case PopupPlacement.BottomLeft:
+      case PopupPlacement.BottomRight:
+      case PopupPlacement.TopLeft:
+        return { marginLeft: -leftFix + arrowOffset };
+      default:
+        return { marginTop: -topFix + arrowOffset };
+    }
+  };
+
+  positionPopup = () => {
+    // `findDOMNode(this)` will search for the DOM node for the current component
+    // first it will find a React.Fragment (see `render`),
+    // so it will get the DOM node of the first child, i.e. DOM node of `this.props.children`
+    // docs: https://reactjs.org/docs/refs-and-the-dom.html#exposing-dom-refs-to-parent-components
+
+    // eslint-disable-next-line react/no-find-dom-node
+    const toggleNode = findDOMNode(this);
+
+    if (toggleNode && toggleNode instanceof Element && this.popupNode.current) {
+      const toggleRect = toggleNode.getBoundingClientRect();
+      const { width, height } = this.popupNode.current.getBoundingClientRect();
+      let left = 0;
+      let top = 0;
+
+      switch (this.getPlacement()) {
+        case PopupPlacement.Bottom:
+          left = toggleRect.left + toggleRect.width / 2 - width / 2;
+          top = toggleRect.top + toggleRect.height;
+          break;
+        case PopupPlacement.BottomLeft:
+          left = toggleRect.left;
+          top = toggleRect.top + toggleRect.height;
+          break;
+        case PopupPlacement.BottomRight:
+          left = toggleRect.left + toggleRect.width - width;
+          top = toggleRect.top + toggleRect.height;
+          break;
+        case PopupPlacement.LeftTop:
+          left = toggleRect.left - width;
+          top = toggleRect.top;
+          break;
+        case PopupPlacement.RightTop:
+          left = toggleRect.left + toggleRect.width;
+          top = toggleRect.top;
+          break;
+        case PopupPlacement.RightBottom:
+          left = toggleRect.left + toggleRect.width;
+          top = toggleRect.top + toggleRect.height - height;
+          break;
+        case PopupPlacement.TopLeft:
+          left = toggleRect.left;
+          top = toggleRect.top - height;
+          break;
+      }
+
+      // save width and height (and later set in `render`) to avoid resizing the popup element,
+      // when it's placed close to the window edge
+      this.setState({
+        left: window.pageXOffset + left,
+        top: window.pageYOffset + top,
+        width,
+        height,
+      });
+    }
+  };
+
+  renderActual = ({ leftFix = 0, topFix = 0 }) => {
+    const { className, overlay, noPadding } = this.props;
+    const placement = this.getPlacement();
+    const style = isMeasured(this.state)
+      ? {
+          left: this.state.left + leftFix,
+          top: this.state.top + topFix,
+          width: this.state.width,
+          height: this.state.height,
+        }
+      : undefined;
+
+    return (
+      <Popup
+        arrowStyle={
+          isMeasured(this.state)
```

Comment:
NP: You're calling this `isMeasured()` twice. You could also just check `style`:

```tsx
        arrowStyle={
         style
            ? this.adjustArrowPosition(placement, { leftFix, topFix })
            : undefined
        }
```